src/iceberg/expression.h

```diff
@@ -0,0 +1,167 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+/// \file iceberg/expression.h
+/// Expression interface for Iceberg table operations.
+
+#include <memory>
+#include <string>
+
+#include "iceberg/expected.h"
+#include "iceberg/iceberg_export.h"
+#include "iceberg/result.h"
+
+namespace iceberg {
+
+/// \brief Type alias for shared pointer to Expression
+class Expression;
+using ExpressionPtr = std::shared_ptr<Expression>;
+
+/// \brief Represents a boolean expression tree.
+class ICEBERG_EXPORT Expression {
+ public:
+  /// Operation types for expressions
+  enum class Operation {
+    kTrue,
+    kFalse,
+    kIsNull,
+    kNotNull,
+    kIsNan,
+    kNotNan,
+    kLt,
+    kLtEq,
+    kGt,
+    kGtEq,
+    kEq,
+    kNotEq,
+    kIn,
+    kNotIn,
+    kNot,
+    kAnd,
+    kOr,
+    kStartsWith,
+    kNotStartsWith,
+    kCount,
+    kCountStar,
+    kMax,
+    kMin
```

Comment:
Do we want to add these right away? I would suggest adding these in a later PR, because:

- `Min`/`Max` can not always be obtained from the metadata, because the metadata is missing, or the value might be truncated (Iceberg applies `truncate[16]` by default on min/max statistics.
- Counts can be problematic to do purely from the metadata when you have merge on read deletes, or it cannot be determined from the metadata that the whole file needs to be included in the scan.

Therefore, I think it is more up to the query engine to implement this kind of operations.

Reply:
Ah, ignore this comment. I see that this has been added to Java as well 👍